presto-elasticsearch/src/main/java/com/facebook/presto/elasticsearch/ElasticsearchQueryBuilder.java

```diff
@@ -0,0 +1,254 @@
+        }
+        scrollTime = config.getScrollTime();
+        scrollSize = config.getScrollSize();
+    }
+
+    public SearchRequestBuilder buildScrollSearchRequest()
+    {
+        StringBuilder shardBuilder = new StringBuilder("_shards:");
+        shardBuilder.append(shard);
+        String indices = index != null && !index.isEmpty() ? index : "_all";
+        SearchRequestBuilder searchRequestBuilder = client.prepareSearch(indices)
+                .setTypes(type)
+                .setSearchType(SearchType.QUERY_AND_FETCH)
+                .setScroll(new TimeValue(scrollTime.toMillis()))
+                .setQuery(buildSearchQuery())
+                .setPreference(shardBuilder.toString())
+                .setSize(scrollSize);
+        log.debug("ElasticSearch Request: " + searchRequestBuilder.toString());
+        return searchRequestBuilder;
+    }
+
+    public SearchScrollRequestBuilder prepareSearchScroll(String scrollId)
+    {
+        return client.prepareSearchScroll(scrollId).setScroll(new TimeValue(scrollTime.toMillis()));
+    }
+
+    private QueryBuilder buildSearchQuery()
+    {
+        BoolQueryBuilder elasticQueryBuilder = new BoolQueryBuilder();
+        for (ElasticsearchColumnHandle column : columns) {
+            Type type = column.getColumnType();
+            tupleDomain.getDomains().ifPresent((domains) -> {
+                Domain domain = domains.get(column);
+                if (domain != null) {
+                    elasticQueryBuilder.must(buildPredicate(column.getColumnJsonPath(), domain, type, column.getIsList()));
+                }
+            });
+        }
+        return elasticQueryBuilder.hasClauses() ? elasticQueryBuilder : new MatchAllQueryBuilder();
+    }
+
+    private QueryBuilder buildPredicate(String columnName, Domain domain, Type type, boolean isList)
+    {
+        checkArgument(domain.getType().isOrderable(), "Domain type must be orderable");
+        BoolQueryBuilder elasticQueryBuilder = new BoolQueryBuilder();
+
+        if (domain.getValues().isNone() && domain.isNullAllowed()) {
+            elasticQueryBuilder.mustNot(new ExistsQueryBuilder(columnName));
+            return elasticQueryBuilder;
+        }
+
+        if (domain.getValues().isAll()) {
+            elasticQueryBuilder.must(new ExistsQueryBuilder(columnName));
+            return elasticQueryBuilder;
+        }
+
+        if (isList) {
+            return buildMatchQuery(elasticQueryBuilder, columnName, domain, type);
+        }
+        return buildTermQuery(elasticQueryBuilder, columnName, domain, type);
+    }
+
+    private QueryBuilder buildMatchQuery(BoolQueryBuilder builder, String columnName, Domain domain, Type type)
+    {
+        List<Object> includeValues = new ArrayList<>();
+        List<Object> excludeValues = new ArrayList<>();
+        for (Range range : domain.getValues().getRanges().getOrderedRanges()) {
+            checkState(!range.isAll());
```

Comment:
please add an error message to `checkState` call. I have seen other `check*` calls without error messages, please add error messages to those too.